libs/base/Data/Vect.idr

```diff
@@ -780,9 +780,15 @@ range {len=S _} = FZ :: map FS range
 ||| transpose [[1,2], [3,4], [5,6], [7,8]]
 ||| ```
 public export
-transpose : {n : _} -> Vect m (Vect n elem) -> Vect n (Vect m elem)
-transpose []        = replicate _ []                -- = [| [] |]
-transpose (x :: xs) = zipWith (::) x (transpose xs) -- = [| x :: xs |]
+transpose : {n : _} -> (1 array : Vect m (Vect n elem)) -> Vect n (Vect m elem)
+transpose []        = replicate _ []          -- = [| [] |]
+transpose (x :: xs) = helper x (transpose xs) -- = [| x :: xs |]
+  where -- Can't use zipWith since it doesn't preserve linearity
+    helper : (1 _ : Vect a elem)
+          -> (1 _ : Vect a (Vect b elem))
+          -> Vect a (Vect (S b) elem)
+    helper [] [] = []
+    helper (x::xs) (tl::tls) = (x::tl) :: helper xs tls
```

Comment:
We cannot use `zipWith` but we could use applicative combinators.
Using Haskell notations, this is `(::) <$> xs <*> tls`.

Reply:
Alternatively we could come up with a naming scheme & provide
linear variants of the higher order functions. e.g. `zipWith1`.

Reply:
Applicative combinators don't work either: `map` only maps unrestricted arrows to unrestricted arrows, and similary for `<*>`.  (I'm also not yet convinced that it *should* work in the general case – that is a question for someone with more knowledge of category theory than me.)

As for having separate linear variants for the higher-order functions, I like the idea, at least until Idris gets multiplicity polymorphism (which I'm hoping it does).  I don't like the `1` suffix though, due to potential confusion with `fold[lr]1`.  However, I think that adding those should be the job of a different PR than this one.